src/Semantic/Git.hs

```diff
@@ -24,22 +30,40 @@ clone url path = sh $ do
 -- | git cat-file -p
 catFile :: FilePath -> OID -> IO Text
 catFile gitDir (OID oid) = sh $ do
-  run "git" [pack ("--git-dir=" <> gitDir), "cat-file", "-p", oid]
+  run "git" ["-C", pack gitDir, "cat-file", "-p", oid]
 
 -- | git ls-tree -rz
 lsTree :: FilePath -> OID -> IO [TreeEntry]
-lsTree gitDir (OID sha) = sh $ do
-  out <- run "git" [pack ("--git-dir=" <> gitDir), "ls-tree", "-rz", sha]
-  pure $ mkEntry <$> splitOn "\NUL" out
-  where
-    mkEntry row | [mode, ty, rest] <- splitOn " " row
-                , [oid, path] <- splitOn "\t" rest
-                = TreeEntry (objectMode mode) (objectType ty) (OID oid) (unpack path)
-                | otherwise = nullTreeEntry
+lsTree gitDir (OID sha) = sh $ parseEntries <$> run "git" ["-C", pack gitDir, "ls-tree", "-rz", sha]
 
 sh :: MonadIO m => Sh a -> m a
 sh = shelly . silently . onCommandHandles (initOutputHandles (`hSetBinaryMode` True))
 
+-- | Parses an list of entries separated by \NUL, and on failure return []
+parseEntries :: Text -> [TreeEntry]
+parseEntries text = case parseOnly everything text of
```

Comment:
I think this can be a good deal simpler. Pull in the `parsers` library and use this.

```
parseOnly ((entryParser `sepEndBy` char '\NUL') <* endOfInput)
```

You shouldn't need to call `feed` or override `parseOnly`.

Reply:
`parseOnly` from Attoparsec? Or from another library? I was using this approach because `Data.Attoparsec.Text.parseOnly` returns minimal error information, which I ran into while debugging my parser. I'm happy to do this, but it will further degrade errors.

Reply:
Remove it for now, and if it becomes a problem we can switch to Megaparsec or Trifecta (which are slower but have way better messaging).